### DISCUSSION TITLE
Error migrating to V4 (Typescript / Immer / produce)

### DISCUSSION THREAD

**Discussion by eddyhdzg** [pmndrs/zustand#1261] [Q&A] [answered]
Updated from  v3 to v4 and now there is a Typescript error.

Current middleware from:
[https://github.com/pmndrs/zustand/blob/How-to-type-immer-middleware-in-TypeScript](https://github.com/pmndrs/zustand/blob/b85f6413f1eed2b8c9f6e6e402c272f4da3b9c2d/docs/recipes/recipes.mdx#middleware:~:text=How%20to%20type%20immer%20middleware%20in%20TypeScript)

```ts
// Zustand v3
const immer =
  <T extends State>(
    config: StateCreator<T, (fn: (draft: Draft<T>) => void) => void>
  ): StateCreator<T> =>
  (set, get, api) =>
    config((fn) => set(produce<T>(fn)), get, api)
```

Any type-safe solutions?

**Comment by dai-shi** [comment, marked as answer]
Typing middleware is very hard. We ended up with providing `immer` middleware from the zustand library.
https://github.com/pmndrs/zustand#immer-middleware

**Reply by eddyhdzg** [reply]
With "zustand": "^4.1.1" it is now working, past versions of the ```immer``` middleware didn't work.
